bpf/kmesh/workload/include/authz.h

```diff
@@ -166,49 +254,143 @@ rule_match_check(Istio__Security__Rule *rule, struct xdp_info *info, struct bpf_
+    res = bpf_map_lookup_elem(&tailcall_info_map, &tuple_key);
+    if (!res) {
+        BPF_LOG(ERR, AUTH, "Failed to retrieve res from map");
+        return XDP_PASS;
     }
 
+    policies = res->policies;
+    if (!policies) {
+        return XDP_PASS;
+    }
+
+    if (res->policy_index < 0 || res->policy_index >= MAX_MEMBER_NUM_PER_POLICY) {
+        BPF_LOG(ERR, AUTH, "Policy index out of bounds");
+        return XDP_PASS;
+    }
+
+    // Safely access policyId and check if the policy exists
+    if (bpf_probe_read_kernel(&policyId, sizeof(policyId), (void *)(policies->policyIds + res->policy_index)) != 0) {
+        BPF_LOG(ERR, AUTH, "Failed to read policyId, thrown it to user auth");
+        if (bpf_map_delete_elem(&tailcall_info_map, &tuple_key) != 0) {
+            BPF_LOG(DEBUG, AUTH, "Failed to delete context from map");
+        }
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+    }
+    policy = map_lookup_authz(policyId);
+    if (!policy) {
+        // if no policy matches in xdp, thrown it to user auth
+        if (bpf_map_delete_elem(&tailcall_info_map, &tuple_key) != 0) {
+            BPF_LOG(DEBUG, AUTH, "Failed to delete context from map");
+        }
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+    } else {
+        rulesPtr = kmesh_get_ptr_val(policy->rules);
+        if (!rulesPtr) {
+            BPF_LOG(ERR, AUTH, "failed to get rules from policy %s\n", kmesh_get_ptr_val(policy->name));
+            return XDP_DROP;
+        }
+        res->rulesPtr = rulesPtr;
+        res->n_rules = policy->n_rules;
+        res->action = policy->action;
+        ret = bpf_map_update_elem(&tailcall_info_map, &tuple_key, res, BPF_ANY);
+        if (ret < 0) {
+            BPF_LOG(ERR, AUTH, "Failed to update map, error: %d", ret);
+            return XDP_DROP;
+        }
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_RULE_CHECK);
+    }
+    return XDP_PASS;
+}
+
+SEC("xdp_auth")
+int rule_check(struct xdp_md *ctx)
+{
+    struct match_ctx *res;
+    struct bpf_sock_tuple tuple_key = {0};
+    struct xdp_info info = {0};
+    void *rulesPtr;
+    __u64 rule_addr;
+    void *rule;
+    int ret;
+    int i;
+
+    if (construct_tuple_key(ctx, &tuple_key, &info) != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "Failed to get tuple key in rule_check");
+        return XDP_ABORTED;
+    }
+
+    res = bpf_map_lookup_elem(&tailcall_info_map, &tuple_key);
+    if (!res) {
+        BPF_LOG(ERR, AUTH, "Failed to retrieve match_ctx from map");
+        return XDP_PASS;
+    }
     for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
-        if (i >= policy->n_rules) {
+        if (i >= res->n_rules) {
+            BPF_LOG(DEBUG, AUTH, "Rule index %d exceeds rule count %d, exiting loop", i, res->n_rules);
             break;
         }
-        rule = (Istio__Security__Rule *)kmesh_get_ptr_val((void *)*((__u64 *)rulesPtr + i));
+        if (!res) {
```

Comment:
redundant check with line 342?

Reply:
> redundant check with line 342?

In xdp type eBPF program, the null pointer check may fail. An additional check  is needed here, otherwise the verifier will report an error.